File: hacking_with_python/missing_security_headers.py
import requests
import concurrent.futures
from termcolor import colored

# Define the list of security headers to check
headers_to_check = ['X-Content-Type-Options', 'X-Frame-Options', 'X-XSS-Protection', 'Content-Security-Policy']

def test_security_headers(url):
    results = {'url': url, 'missing_headers': []}
    
    try:
        response = requests.get(url, timeout=5)  # Set a timeout for the HTTP request
        
        if response.status_code == 200:
            print(f"Testing {url}...")
            for header in headers_to_check:
                if header not in response.headers:
                    results['missing_headers'].append(header)
            return results
        else:
            print(f"Failed to retrieve {url}: Status code {response.status_code}")
            return None
    
    except Exception as e:
        print(f"Error while testing {url}: {e}")
        return None

def print_results(results):
    if results:
        print("\n" + "=" * 50)
        print(f"Results for {results['url']}:\n")
        if results['missing_headers']:
            print(colored("Missing Security Headers:", "red"))
            for header in results['missing_headers']:
                print(colored(f"- {header}", "red"))
            print("\nVulnerability Exploitation:")
            print("Attackers could exploit this vulnerability to perform various attacks such as clickjacking, XSS, etc.")
        else:
            print(colored("All Security Headers Found!", "green"))
            print("The website is protected against the Missing Security Headers vulnerability.")
        print("=" * 50 + "\n")

if __name__ == "__main__":
    target_url = input("Enter your target website URL: ")
    
    with concurrent.futures.ThreadPoolExecutor() as executor:
        future_results = [executor.submit(test_security_headers, target_url) for _ in range(5)]  # Send 5 concurrent requests
    
        for future in concurrent.futures.as_completed(future_results):
            print_results(future.result())

File: hacking_with_c++/missing_security_headers.cpp
#include <iostream>
#include <string>
#include <curl/curl.h>

// Colors for console output
#define RED "\033[1;31m"
#define GREEN "\033[1;32m"
#define YELLOW "\033[1;33m"
#define RESET "\033[0m"

// Function to perform HTTP GET request using libcurl
std::string performHttpGet(const std::string& url) {
    CURL *curl;
    CURLcode res;
    std::string response;

    curl = curl_easy_init();
    if(curl) {
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, [](void *buffer, size_t size, size_t nmemb, void *userp) -> size_t {
            ((std::string*)userp)->append((char*)buffer, size * nmemb);
            return size * nmemb;
        });
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);

        res = curl_easy_perform(curl);
        curl_easy_cleanup(curl);

        if(res != CURLE_OK) {
            std::cerr << RED << "Failed to perform HTTP GET request: " << curl_easy_strerror(res) << RESET << std::endl;
        }
    }

    return response;
}

// Function to check for missing security headers
void checkSecurityHeaders(const std::string& url) {
    std::string response = performHttpGet(url);

    // Check for missing security headers
    bool missingHeaders = false;
    if (response.find("X-Content-Type-Options") == std::string::npos) {
        std::cout << YELLOW << "Missing X-Content-Type-Options header" << RESET << std::endl;
        missingHeaders = true;
    }
    if (response.find("X-Frame-Options") == std::string::npos) {
        std::cout << YELLOW << "Missing X-Frame-Options header" << RESET << std::endl;
        missingHeaders = true;
    }
    if (response.find("X-XSS-Protection") == std::string::npos) {
        std::cout << YELLOW << "Missing X-XSS-Protection header" << RESET << std::endl;
        missingHeaders = true;
    }

    if (!missingHeaders) {
        std::cout << GREEN << "No missing security headers found" << RESET << std::endl;
    }

    // Information on how to exploit the vulnerability
    std::cout << "To exploit this vulnerability, an attacker could inject malicious content or perform clickjacking attacks." << std::endl;
}

int main() {
    std::string targetWebsite;

    std::cout << "Enter your target website URL: ";
    std::cin >> targetWebsite;

    // Check for missing security headers
    std::cout << "Checking for missing security headers on: " << targetWebsite << std::endl;
    checkSecurityHeaders(targetWebsite);

    return 0;
}
